code/Chapter5/test_dependency_injection.cpp

```diff
@@ -0,0 +1,84 @@
+#include <iostream>
+#include <cmath>
+#include "gtest/gtest.h"
+
+/// @brief A "complex" calculation of the norm.
+/// @param array The vector to calculate the norm for.
+/// @return The actual norm of the vector.
+double calculate_norm(std::vector<int> array)
+{
+  double output{0};
+  for (int i{0}; i < array.size(); ++i)
+  {
+    output = array[i] * array[i];
+  }
+  return std::sqrt(output);
+}
+
+/// @brief Just a stub, so we can get a simple value to test things with.
+/// @param array The vector to calculate the norm for.
+/// @return The norm, a dummy value in this case.
+double norm_stub(std::vector<int> array)
+{
+  return 10.0;
+}
+
+/// @brief We cannot do dependency injection, so the result depends on what calculate_norm is doing
+/// @param array The vector to normalize.
+void normalize_v1(std::vector<int> array)
+{
+  double norm{calculate_norm(array)};
+
+  for (int i{0}; i < array.size(); ++i)
+  {
+    array[i] /= norm;
+  }
+}
+
+/// @brief The norm calculation is injected, so we control that value.
+/// @param array The vector to normalize.
+/// @param func The function used to calculate the norm.
+void normalize_v2(
+    std::vector<int> array,
+    std::function<double(std::vector<int>)> func = calculate_norm)
+{
+  double norm{func(array)};
+
+  for (int i{0}; i < array.size(); ++i)
+  {
+    array[i] /= norm;
+  }
+}
+
+TEST(NormalizeTest, WithoutDependencyInjection)
+{
+  std::vector<int> input{1, 2, 3};
+
+  // We need the exact value of the norm in order to check that
+  // the calculation is correct.
+  double factor{calculate_norm(input)};
+  std::vector<double> copy{1, 2, 3};
+
+  normalize_v2(input);
+
+  for (int i{0}; i < input.size(); ++i)
+  {
+    EXPECT_EQ(input[i] / factor, copy[i] / factor);
+  }
+}
```

Comment:
I tried to run these tests and they all pass but I think there is small issue here.
1. In `normalize_v2(input)`, the array gets divided by the norm.
2. While checking you should not need to divide the array again by factor (`input[i]/factor`) which is storing the norm of the array. 
3. I am not sure if you have intentionally chosen to pass the vector by value or it is a mistake. At present, the changes inside the function will not be reflected outside. This is why, you had to divide it by factor outside the function. In this scenario, the normalise_function v2 is not doing anything and only reason your test passes is because of the `EXPECT_EQ(input[i] / factor, copy[i] / factor)` which is doing the same thing on 2 copies.

Reply:
Well, is not a small issue: my test is plainly wrong because my code is wrong 😅 . Thanks for pointing this out.

Reply:
Fixed in 5337018

I've rewritten the code to pass the input vectors by reference and as constants, and to return the normalised array rather than doing it in place. That's more functional and cleaner, I feel, than modifying things in place.